drivers/dma/pl330dma/pl330.c

```diff
@@ -0,0 +1,502 @@
+#include "driver.h"
+#include "stdint.h"
+
+#define bool int
+#define MS_IN_US 1000
+
+static ULONG Pl330DmaDebugLevel = 100;
+static ULONG Pl330DmaDebugCatagories = DBG_INIT || DBG_PNP || DBG_IOCTL;
+
+void udelay(ULONG usec) {
+	LARGE_INTEGER Interval;
+	Interval.QuadPart = -10 * (LONGLONG)usec;
+	KeDelayExecutionThread(KernelMode, FALSE, &Interval);
+}
+
+UINT32 read32(PPL330DMA_CONTEXT pDevice, UINT32 reg)
+{
+	return *(UINT32 *)((CHAR *)pDevice->MMIOAddress + reg);
```

Comment:
This makes me nervous. This implementation gives the compiler freedom to perform the reads and writes out-of-order from what you wrote in your code, and to reorder other memory updates with the register read.

WDM provides the READ_REGISTER_ULONG, which ensures that things happen in the order you meant them to happen. Consider using that instead.